### PR TITLE
Mz/noise level check

### PR DESCRIPTION
**Pull request by mayeul-zama**

### CODE REVIEW

**Comment by mayeul-zama** [comment]
@slab-ci cpu_fast_test

File: tfhe/src/integer/server_key/crt/add_crt.rs [hunk 1]
```diff
@@ -31,32 +32,34 @@ impl ServerKey {
         ct_left: &mut CrtCiphertext,
         ct_right: &mut CrtCiphertext,
     ) -> CrtCiphertext {
-        if !self.is_crt_add_possible(ct_left, ct_right) {
+        if self.is_crt_add_possible(ct_left, ct_right).is_err() {
             self.full_extract_message_assign(ct_left);
             self.full_extract_message_assign(ct_right);
         }
-        assert!(self.is_crt_add_possible(ct_left, ct_right));
+        assert!(self.is_crt_add_possible(ct_left, ct_right).is_ok());
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/crt/add_crt.rs:39]: maybe unwrap to get the error message (same remark everywhere)

File: tfhe/src/shortint/ciphertext/mod.rs [hunk 2]
```diff
@@ -13,6 +13,20 @@ use super::parameters::{CiphertextConformanceParams, CiphertextListConformancePa
 #[derive(Debug, PartialEq, Eq, Copy, Clone, Serialize, Deserialize)]
 pub struct Degree(pub usize);
 
+/// This tracks the amount of noise in a ciphertext.
```
Comment (IceTDrinker) [inline tfhe/src/shortint/ciphertext/mod.rs:16]: docstring does not look right, the maximum amount of noise allowed in a [`Ciphertext`] maybe ?

File: tfhe/src/shortint/engine/server_side/mod.rs [hunk 3]
```diff
@@ -159,13 +160,19 @@ impl ShortintEngine {
             &mut self.encryption_generator,
         );
 
+        let max_noise_level = MaxNoiseLevel::new(divide_ceil(
+            cks.parameters.carry_modulus().0 * cks.parameters.message_modulus().0 - 1,
+            cks.parameters.message_modulus().0 - 1,
+        ));
```
Comment (IceTDrinker) [inline tfhe/src/shortint/engine/server_side/mod.rs:166]: i would probably use a plain divide (which floors) so that we are sure to respect the noise ? any reason to prefer the ceil ? (I believe all our current parameter sets should have round divisions for the noise)

File: tfhe/src/shortint/engine/wopbs/mod.rs [hunk 4]
```diff
@@ -155,16 +156,27 @@ impl ShortintEngine {
             &mut self.encryption_generator,
         );
 
+        let max_noise_level = MaxNoiseLevel::new(divide_ceil(
+            parameters.carry_modulus.0 * parameters.message_modulus.0 - 1,
+            parameters.message_modulus.0 - 1,
+        ));
+
```
Comment (IceTDrinker) [inline tfhe/src/shortint/engine/wopbs/mod.rs:163]: same remark for the noise

maybe rename this one to indicate it's the noise for the wopbs server key

File: tfhe/src/shortint/engine/wopbs/mod.rs [hunk 5]
```diff
@@ -155,16 +156,27 @@ impl ShortintEngine {
...
             bootstrapping_key: ShortintBootstrappingKey::Classic(small_bsk),
             message_modulus: parameters.message_modulus,
             carry_modulus: parameters.carry_modulus,
             max_degree: MaxDegree(parameters.message_modulus.0 * parameters.carry_modulus.0 - 1),
+            max_noise_level,
             ciphertext_modulus: parameters.ciphertext_modulus,
             pbs_order: cks.parameters.encryption_key_choice().into(),
         };
 
+        let max_noise_level = MaxNoiseLevel::new(divide_ceil(
```
Comment (IceTDrinker) [inline tfhe/src/shortint/engine/wopbs/mod.rs:175]: rename to indicate it's the max noise for the shortint server key (non wop)

File: tfhe/src/shortint/server_key/mod.rs [hunk 6]
```diff
@@ -55,6 +57,9 @@ impl Display for CheckError {
             Self::CarryFull => {
                 write!(f, "The carry buffer is full")
             }
+            Self::NoiseTooBig => {
```
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/mod.rs:60]: do we maybe want to store the noise level requested and the max noise level and format that in the error message ?

Same maybe for the carry full above, unsure

It's not necessary to go forward, could be nice potentially

File: tfhe/src/shortint/server_key/mod.rs [hunk 7]
```diff
@@ -951,12 +991,18 @@ impl From<CompressedServerKey> for ServerKey {
             },
         );
 
+        let max_noise_level = MaxNoiseLevel::new(divide_ceil(
+            carry_modulus.0 * message_modulus.0 - 1,
+            message_modulus.0 - 1,
+        ));
```
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/mod.rs:997]: I think we need a formula, the code is repeated 4 or 5 times, it should take a carry modulus and a message modulus

It will allow to add some small tests for e.g. 2_2 like values and check we get a max noise level of 5

**Review by IceTDrinker — COMMENTED** [review]
Thanks for what looks like tedious work, some comments here and there, commits need squashing I believe

File: tfhe/src/integer/server_key/crt/add_crt.rs (hunk 1 above)
Comment (mayeul-zama) [reply tfhe/src/integer/server_key/crt/add_crt.rs:39]: Sure
If we could enable clippy lints on doc tests, we would probably have many such warnings

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/crt/add_crt.rs:39]: there is an old issue in cargo for that, but if the doctests prove to be broken we'll likely add something to lint doctests

File: tfhe/src/shortint/ciphertext/mod.rs (hunk 2 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/ciphertext/mod.rs:16]: Fixed

File: tfhe/src/shortint/engine/server_side/mod.rs (hunk 3 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/mod.rs:166]: Indeed, replaced by a integer divide

File: tfhe/src/shortint/server_key/mod.rs (hunk 7 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/mod.rs:997]: Put into `MaxNoiseLevel::from_msg_carry_modulus`

File: tfhe/src/shortint/engine/wopbs/mod.rs (hunk 4 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/wopbs/mod.rs]: Done

File: tfhe/src/shortint/engine/wopbs/mod.rs (hunk 5 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/engine/wopbs/mod.rs]: Done

File: tfhe/src/shortint/server_key/mod.rs (hunk 6 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/mod.rs]: Done both

File: tfhe/src/shortint/engine/server_side/mul.rs [hunk 8]
```diff
@@ -132,7 +137,7 @@ impl ShortintEngine {
                     self.message_extract_assign(server_key, ct_right);
                 }
             }
-            assert!(server_key.is_mul_possible(ct_left, ct_right));
+            server_key.is_mul_possible(ct_left, ct_right).unwrap();
```
Comment (tmontaigu) [inline tfhe/src/shortint/engine/server_side/mul.rs:140]: Sometimes you keep asserts, sometimes you use.unwrap(). I think it should be consistent

And I would personally prefer assert, as to me it makes the intent clearer that

Comment (tmontaigu) [reply tfhe/src/shortint/engine/server_side/mul.rs:140]: hum yeah we would not have the info on whether its a noise or degree issue 🤔

Comment (mayeul-zama) [reply tfhe/src/shortint/engine/server_side/mul.rs:140]: I changed the return type of `is_mul_possible` (and similar functions) from `bool` to `Result<(), CheckError>` to know if it is a degree or a noise error.
There should not be any `assert!(...is_ok())`

Comment (tmontaigu) [reply tfhe/src/shortint/engine/server_side/mul.rs:140]: yeah that is fine

File: tfhe/src/core_crypto/commons/ciphertext_modulus.rs [hunk 9]
```diff
@@ -298,10 +298,8 @@ mod tests {
         assert!(std::mem::align_of::<CiphertextModulus<u128>>() == std::mem::align_of::<u128>());
 
         {
-            let mod_32_res = CiphertextModulus::<u32>::try_new_power_of_2(32);
-            assert!(mod_32_res.is_ok());
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/commons/ciphertext_modulus.rs:302]: why was this modified ? it has nothing to do with the current PR right ?

Comment (mayeul-zama) [reply tfhe/src/core_crypto/commons/ciphertext_modulus.rs:302]: I did this change while removing `assert!(*.is_ok());`s from another version of the PR
I think it's better this way
I'll put that it a separate commit

Comment (IceTDrinker) [reply tfhe/src/core_crypto/commons/ciphertext_modulus.rs:302]: I would prefer if we avoid such changes which are unrelated to the ongoing work, it makes the PRs more noisy and delays the reviews :)

File: tfhe/src/integer/server_key/radix/add.rs [hunk 10]
```diff
@@ -119,17 +119,34 @@ impl ServerKey {
...
+                        left_block.message_modulus.0 * left_block.carry_modulus.0 - 1,
+                    ),
+                });
             }
+
+            let extracted_carry_noise_level = if first_block {
+                NoiseLevel::ZERO
+            } else {
+                NoiseLevel::NOMINAL
+            };
```
Comment (tmontaigu) [inline tfhe/src/integer/server_key/radix/add.rs:134]: Or init `extracted_carry_noise_level`  to zero before the loop and at the end of the loop set it to  Nominal

File: tfhe/src/integer/server_key/radix/bitwise_op.rs [hunk 11]
```diff
@@ -207,23 +198,33 @@ impl ServerKey {
...
+            .is_functional_bivariate_pbs_possible(ct_left, ct_right)
+            .is_err()
+        {
             self.full_propagate(ct_left);
             self.full_propagate(ct_right);
         }
-        assert!(self.is_functional_bivariate_pbs_possible(ct_left, ct_right));
+        assert!(self
+            .is_functional_bivariate_pbs_possible(ct_left, ct_right)
+            .is_ok());
```
Comment (tmontaigu) [inline tfhe/src/integer/server_key/radix/bitwise_op.rs:210]: Shouldn't these be unwrap ?

File: tfhe/src/integer/server_key/radix/sub.rs [hunk 12]
```diff
@@ -136,19 +137,32 @@ impl ServerKey {
...
+                    degree: Degree(degree_after_add + preceding_block_carry),
+                    max_degree: MaxDegree(total_modulus - 1),
+                });
             }
 
+            let extracted_carry_noise_level = if first_block {
+                NoiseLevel::ZERO
+            } else {
+                NoiseLevel::NOMINAL
+            };
```
Comment (tmontaigu) [inline tfhe/src/integer/server_key/radix/sub.rs:150]: same as for addition, init `extracted_carry_noise_level` to zero before the loop then assign Nominal to it

File: tfhe/src/integer/server_key/radix/bitwise_op.rs (hunk 11 above)
Comment (mayeul-zama) [reply tfhe/src/integer/server_key/radix/bitwise_op.rs:210]: Yes, thanks!
I guess my regex did not allow newlines :)

File: tfhe/src/integer/server_key/radix/sub.rs (hunk 12 above)
Comment (mayeul-zama) [reply tfhe/src/integer/server_key/radix/sub.rs]: Done

File: tfhe/src/integer/server_key/radix/bitwise_op.rs (hunk 11 above)
Comment (mayeul-zama) [reply tfhe/src/integer/server_key/radix/bitwise_op.rs]: Done

File: tfhe/src/integer/server_key/radix/add.rs (hunk 10 above)
Comment (mayeul-zama) [reply tfhe/src/integer/server_key/radix/add.rs]: Done

File: tfhe/src/core_crypto/commons/ciphertext_modulus.rs (hunk 9 above)
Comment (mayeul-zama) [reply tfhe/src/core_crypto/commons/ciphertext_modulus.rs]: Done

File: tfhe/src/shortint/ciphertext/mod.rs [hunk 13]
```diff
@@ -8,11 +8,49 @@ use std::cmp;
...
+
+    pub fn get(&self) -> usize {
+        self.0
+    }
+
+    // This function is valid for current parameters as they guarantee the p-error for a norm2 noise
+    // limit equal to the norm2 limit which guarantees a clean padding bit
+    //
+    // TODO: remove this functions once noise norm2 constraint is decorrelated and stored in
+    // parameter sets
```
Comment (IceTDrinker) [inline tfhe/src/shortint/ciphertext/mod.rs:35]: if we don't have an issue for this let's create one and assign yourself on it for now

**Review by IceTDrinker — COMMENTED** [review]
I think we are good to go I'll wait for Thomas to do a last pass to avoid having change requests while tests are running.

Thansk a lot !

**Review by tmontaigu — APPROVED** [review]

File: tfhe/src/shortint/ciphertext/mod.rs (hunk 13 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/ciphertext/mod.rs:35]: https://github.com/zama-ai/tfhe-rs-internal/issues/339
